### PR TITLE
Enable signing without private key

### LINKED ISSUE #282: CSR with PublicKey only
Usecase:

* Client generates keypair
* Sends public key to server (in form of CSR, or CSR is generated on server itself)
* Server signs him a client cert with CA existing only on server
* Returns this cert back to client

Mb I'm missing something, but didn't find a way to do it. The only possible flow now is:

* Server generates both private and public key for client (`KeyPair::generate`)
* Signs a cert with existing CA
* Returns both cert and private key to a client.

### PR DESCRIPTION
**Pull request by djc**
Takes #288 and addresses the comments.

Fixes #282.

### CODE REVIEW

File: rcgen/src/error.rs [hunk 1]
```diff
@@ -91,6 +94,8 @@ impl fmt::Display for Error {
 			)?,
 			#[cfg(not(feature = "crypto"))]
 			MissingSerialNumber => write!(f, "A serial number must be specified")?,
+			#[cfg(feature = "x509-parser")]
+			X509(e) => write!(f, "X509 error: {e}")?,
```
Comment (cpu) [inline rcgen/src/error.rs:98]: nit: maybe `X.509 parsing error:` as the prefix?

File: rcgen/src/key_pair.rs [hunk 2]
```diff
@@ -690,6 +690,69 @@ impl<T> ExternalError<T> for Result<T, pem::PemError> {
 	}
 }
 
+/// A public key
+#[derive(Debug)]
```
Comment (cpu) [inline rcgen/src/key_pair.rs:694]: It feels like deriving `Eq` and `PartialEq` could be helpful here. WDYT?

File: rcgen/src/key_pair.rs [hunk 3]
```diff
@@ -690,6 +690,69 @@ impl<T> ExternalError<T> for Result<T, pem::PemError> {
...
+	pub(crate) alg: &'static SignatureAlgorithm,
+	pub(crate) subject_public_key: Vec<u8>,
+}
+
+impl SubjectPublicKeyInfo {
+	/// Create a `SubjectPublicKey` value from a PEM-encoded SubjectPublicKeyInfo string
+	#[cfg(all(feature = "x509-parser", feature = "pem"))]
+	pub fn from_pem(pem_str: &str) -> Result<Self, Error> {
+		let spki_der = pem::parse(pem_str)._err()?.into_contents();
+		Self::from_der(&spki_der)
```
Comment (cpu) [inline rcgen/src/key_pair.rs:705]: nit: inline `spki_der`?:

```rust
        Self::from_der(&pem::parse(pem_str)._err()?.into_contents())
```

File: rcgen/src/key_pair.rs [hunk 4]
```diff
@@ -716,6 +779,27 @@ mod test {
 		signature::{EcdsaKeyPair, ECDSA_P256_SHA256_FIXED_SIGNING},
 	};
 
+	#[cfg(all(feature = "x509-parser", feature = "pem"))]
+	#[test]
+	fn test_subject_public_key_parsing() {
+		// NOTE: the other algorithms supported by this crate don't support keygen
```
Comment (cpu) [inline rcgen/src/key_pair.rs:785]: Is this comment accurate? We support RSA and ECDSA P521 keygen w/ aws-lc-rs no?

**Review by cpu — APPROVED** [review]
Overall looks good, thanks for picking this up.

**Comment by kwantam** [comment]
Thanks for doing this and apologies that I have been buried!

**Pull request merged**